### PR TITLE
Add support for EAR deployments

### PR DESCRIPTION
**Pull request by shrishrirang**
- This commit adds support for deploying EAR files.
- Also, new convention is introduced for deployments. Specifically, this is the new deployment order:
   1. /home/site/wwwwroot/app.ear
   2. /home/site/wwwroot/app.war
   3. Apps in /home/site/wwwroot/webapps/*

### CODE REVIEW

**Comment by shrishrirang** [comment]
@chunye

File: shared/appservice/init_container.sh [hunk 1]
```diff
@@ -103,6 +89,34 @@ echo ***Waiting for admin server to be ready
...
+    echo "Found app.ear in /home/site/wwwroot, deploying it"
+    ln -s /home/site/wwwroot/app.ear /tmp/wildfly/appservice/ROOT.ear
+    $JBOSS_HOME/bin/jboss-cli.sh -c "deploy /tmp/wildfly/appservice/ROOT.ear"
+elif [ -f /home/site/wwwroot/app.war ]
+then
+    echo "Found app.war in /home/site/wwwroot, deploying it"
+    ln -s /home/site/wwwroot/app.war /tmp/wildfly/appservice/ROOT.war
+    $JBOSS_HOME/bin/jboss-cli.sh -c "deploy /tmp/wildfly/appservice/ROOT.war"
+else
+    echo "NOT found app.ear or app.ear in /home/site/wwwroot, deploying apps in /home/site/wwwroot/webapps if any"
```
Comment (chunye) [inline shared/appservice/init_container.sh]: app.ear and app.war

**Review by chunye — APPROVED** [review]
Looks good.  Just fix the typo in the comment.

**Comment by stuartraetaylor** [comment]
It would be great if this could also deploy multiple WAR and/or EAR archives.

**Comment by stuartraetaylor** [comment]
Just to add some more details for my use case - I have an app with three WAR files and one EAR file, which is currently deployed to a single Wildfly app service container.

My current approach is to deploy WAR files using war deploy in the usual way, and deploy EAR files using ZIP/FTP deploy to some directory, e.g., `/home/site/wwwroot/archives`.

To support the EAR/WAR archive deployment on startup of the container I have modified startup.sh to deploy the archives based on the existing approach used by init_container.sh -

```
# Deploy archived WARs and EARs on start up.
# see https://github.com/Azure-App-Service/wildfly/blob/dev/shared/misc/init_container.sh#L66
for f in /home/site/wwwroot/archives/*; do
    if [[ -f "$f" ]]; then
        echo ***Copying $f to $JBOSS_HOME/standalone/deployments/
        cp $f $JBOSS_HOME/standalone/deployments/

        markerfile=$JBOSS_HOME/standalone/deployments/$(basename $f).dodeploy

        echo ***Creating marker file $markerfile
        touch $markerfile
    elif [[ -d "$f" ]]; then
        echo "Skipping directory: $f"
    fi
done
```

**Comment by shrishrirang** [comment]
@stuartraetaylor, thanks for sharing your use case! This is the first baby step towards supporting EAR deployments, so this PR doesn't yet add multiple EAR support. We are currently having discussions on the next steps and also regarding the tooling support. Tagging @jvano @JasonFreeberg for sharing your use case with them. 

Btw, is there a reason you prefer multiple war and ear files per Azure web app versus having a 1:1 mapping between Azure web app and the war/ear?

**Comment by shrishrirang** [comment]
@stuartraetaylor : I am merging this PR, but I'll be watching out this thread for your comments on your reasons for preferring a 1:1 mapping. Thanks!

**Comment by stuartraetaylor** [comment]
@shrishrirang thanks for the explanation that makes sense.

In my case the deployment is part of a very early migration of an existing app to Azure services. There still exists some coupling between the WARs and EAR that would need to be addressed before they can be deployed separately. Ideally I think we would have a 1:1 mapping between app service and WAR/EAR.

**Comment by shrishrirang** [comment]
Got it, thanks!

**Comment by selvasingh** [comment]
@stuartraetaylor - thank you for sharing your use case!!

**Pull request merged**